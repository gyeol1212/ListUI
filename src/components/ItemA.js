import React, { useEffect, useRef } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import DefaultItemComponent from './ItemDetail';

const ItemContainer = styled.div`
  width: ${props => `${props.width || '18rem'}`};
  height: ${props => `${props.height || '18rem'}`};
  box-sizing: border-box;
  display: inline-block;
`;

const ItemA = props => {
  const {
    item,
    isSelected,
    index,
    onClickItem,
    itemStyle,
    itemCssObject,
    itemComponent
  } = props;

  const divEl = useRef(null);

  useEffect(() => {
    // 이 component가 선택 되었을 때,
    if (isSelected) {
      // content 한 개의 width
      const contentWidth = divEl.current.offsetWidth;

      // 전체 carousel의 왼쪽 좌표
      const carouselLeft = divEl.current.parentNode.clientLeft;

      // carousel의 왼쪽 좌표 + ( content의 width X content 갯수 )
      divEl.current.parentNode.scrollLeft = carouselLeft + contentWidth * index;
    }
  }, [isSelected, index]);

  return (
    <ItemContainer
      ref={divEl}
      width={itemStyle && itemStyle.width}
      height={itemStyle && itemStyle.height}
      onClick={onClickItem}
      style={itemCssObject}
    >
      {/* <ItemDetail index={index} /> */}
      {itemComponent ? (
        // TODO : index 삭제
        React.cloneElement(itemComponent, {
          index,
          ...itemComponent.props,
          ...item
        })
      ) : (
        <DefaultItemComponent index={index} />
      )}
    </ItemContainer>
  );
};

ItemA.propTypes = {
  item: PropTypes.object,
  index: PropTypes.number,
  isSelected: PropTypes.bool,
  itemStyle: PropTypes.object,
  onClickItem: PropTypes.func,
  itemCssObject: PropTypes.object,
  itemComponent: PropTypes.element
};

export default ItemA;
