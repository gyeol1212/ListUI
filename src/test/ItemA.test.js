import React from 'react';
import { mount } from 'enzyme';

import ItemA from '../components/ItemA';

describe('Item A', () => {
  let wrapper;

  it('renders Item A', () => {
    wrapper = mount(<ItemA />);
    expect(wrapper).toHaveLength(1);
  });

  it('match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
