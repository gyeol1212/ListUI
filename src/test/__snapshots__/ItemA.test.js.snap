// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item A match snapshot 1`] = `
<ItemA>
  <styled.div>
    <div
      className="sc-AykKG fmSQbP"
    >
      <ContentDetail>
        <styled.div>
          <div
            className="sc-AykKC ctRmSy"
          >
            <styled.div>
              <div
                className="sc-AykKD cCRaZj"
              />
            </styled.div>
            <Styled(styled.div)>
              <div
                className="sc-AykKD sc-AykKE jTrkiM"
              />
            </Styled(styled.div)>
            <Styled(styled.div)>
              <div
                className="sc-AykKD sc-AykKF frZAto"
              />
            </Styled(styled.div)>
          </div>
        </styled.div>
      </ContentDetail>
    </div>
  </styled.div>
</ItemA>
`;
